File: 03_DeleteNode/3.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def deleteNode(head, x):
    if not head:
        return None
    
    if x == 1:
        head = head.next
        return head
    
    i = 1
    temp = head
    while i < x - 1:
        i += 1
        temp = temp.next
        
    node = temp.next
    temp.next = node.next
    node = None
    
    return head

def createLinkedList(arr):
    if not arr:
        return None
    head = Node(arr[0])
    current = head
    for val in arr[1:]:
        current.next = Node(val)
        current = current.next
    return head

def printLinkedList(head):
    current = head
    while current:
        print(current.data, end=" -> ")
        current = current.next
    print("None")

if __name__ == "__main__":
    arr = [10, 20, 30, 40, 50]
    head = createLinkedList(arr)
    print("Original Linked List:")
    printLinkedList(head)

    x = 3
    head = deleteNode(head, x)
    print("\nLinked List After Deleting Node", x, ":")
    printLinkedList(head)

File: 03_DeleteNode/3.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* createNode(int data) {
    struct Node* newnode = (struct Node*)malloc(sizeof(struct Node));
    newnode->data = data;
    newnode->next = NULL;
    return newnode;
}

struct Node* deleteNode(struct Node* head, int x) {
    if (head == NULL)
        return NULL;

    if (x == 1) {
        struct Node* temp = head;
        head = head->next;
        free(temp);
        return head;
    }

    struct Node* temp = head;
    int i = 1;

    while (i < x - 1 && temp != NULL) {
        temp = temp->next;
        i++;
    }

    if (temp == NULL || temp->next == NULL)
        return head;

    struct Node* nodeToDelete = temp->next;
    temp->next = nodeToDelete->next;
    free(nodeToDelete);

    return head;
}

void printList(struct Node* head) {
    struct Node* temp = head;
    while (temp != NULL) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

struct Node* createLinkedList(int arr[], int n) {
    if (n == 0)
        return NULL;

    struct Node* head = createNode(arr[0]);
    struct Node* current = head;

    for (int i = 1; i < n; i++) {
        current->next = createNode(arr[i]);
        current = current->next;
    }

    return head;
}

int main() {
    int arr[] = {10, 20, 30, 40, 50};
    int n = sizeof(arr) / sizeof(arr[0]);

    struct Node* head = createLinkedList(arr, n);

    printf("Original Linked List:\n");
    printList(head);

    int x = 3; 
    head = deleteNode(head, x);

    printf("\nLinked List After Deleting Node %d:\n", x);
    printList(head);

    return 0;
}